app/api/business/brief_overview_business.py

```diff
@@ -1,8 +1,69 @@
 from flask_login import current_user
+
 from app.api.business.agreement_business import use_old_work_order_creator
+from app.api.helpers import abort
+from app.api.services import (audit_service, audit_types,
+                              brief_responses_service, briefs)
+from app.emails import send_opportunity_closed_early_email
+from app.tasks import publish_tasks
+from app.tasks.s3 import create_responses_zip
 from app.validation import get_sections as get_validation_sections
 
 
+def can_close_opportunity_early(brief):
+    seller_selector = brief.data.get('sellerSelector', '')
+    invited_sellers = brief.data.get('sellers', {}).keys()
+    number_of_candidates = int(brief.data.get('numberOfSuppliers', 0))
+
+    if brief.status == 'live' and len(invited_sellers) == 1 and (
+        (brief.lot.slug in ['rfx', 'training2'] and seller_selector == 'oneSeller') or
+        (brief.lot.slug == 'specialist' and seller_selector == 'someSellers')
+    ):
+        invited_seller_code = int(invited_sellers.pop())
+        response_count = brief_responses_service.find(
+            brief_id=brief.id,
+            supplier_code=invited_seller_code,
+            withdrawn_at=None
+        ).count()
+
+        if brief.lot.slug in ['rfx', 'training2'] and response_count == 1:
+            return True
+
+        if brief.lot.slug == 'specialist' and response_count == number_of_candidates:
+            return True
+
+    return False
+
+
+def close_opportunity_early(brief_id):
+    brief = briefs.get(brief_id)
+    if not can_close_opportunity_early(brief):
+        abort('Unable to close opportunity {}'.format(brief_id))
```

Comment:
`abort` is a HTTP specific outcome, and this business logic may be used in a non HTTP context. This could return a `False` here or perhaps a specific exception, which the HTTP view then handles and responds with an `abort`.

Reply:
Thanks, this prompted me to rediscover the errors we define in `app/api/business/errors.py`.  I'd completely forgotten about them!  I now raise a few errors which are caught by the view.